packages/contracts-periphery/contracts/universal/op-nft/Optimist.sol

```diff
@@ -0,0 +1,151 @@
+    }
+
+    /**
+     * @notice  Mint the Optimist token.
+     * @dev     Mints the Optimist token to the give recipient address.
+     *          Limits the number of tokens that can be minted to one per address.
+     *          The tokenId is the uint256 of the recipient address.
+     * @param   _recipient  The address of the token recipient.
+     */
+    function mint(address _recipient) public {
+        require(balanceOf(_recipient) == 0, "ALREADY_MINTED");
+        require(isWhitelisted(_recipient), "NOT_WHITELISTED");
+        uint256 tokenId = uint256(uint160(_recipient));
+        _safeMint(_recipient, tokenId);
+    }
+
+    /**
+     * @notice Returns the URI for the token metadata.
+     * @dev The token URI will be stored at baseURI + '/' + tokenId + .json
+     * @param tokenId The token ID to query.
+     * @return The URI for the given token ID.
+     */
+    function tokenURI(uint256 tokenId) public view virtual override returns (string memory) {
+        if (ownerOf(tokenId) == address(0)) {
+            revert("Optimist:::tokenURI: TOKEN_URI_DNE");
+        }
+        return
+            string(
+                abi.encodePacked(
+                    baseURI(),
+                    "/",
+                    Strings.toHexString(tokenId, _ADDRESS_LENGTH),
+                    ".json"
+                )
+            );
+    }
+
+    /**
+     * @notice  Returns whether an address is whitelisted
+     * @dev     The whitelist is an attestation by the admin of this contract
+     * @return  boolean  Whether the address is whitelisted
+     */
+    function isWhitelisted(address _recipient) public view returns (bool) {
+        return
+            attestationStation
+                .attestations(owner(), _recipient, bytes32("optimist.can-mint"))
+                .length > 0;
+    }
+
+    /**
+     * @notice  Returns whether an address is whitelisted
+     * @dev     The whitelist is an attestation by the admin of this contract
+     * @return  boolean  Whether the address is whitelisted
+     */
+    function tokenIdOfOwner(address _owner) public view returns (uint256) {
+        require(balanceOf(_owner) > 0, "Optimist::tokenIdOfOwner: NOT_MINTED");
+        return uint256(uint160(_owner));
+    }
+
+    /**
+     * @notice  (Internal) Optimist Token Base URI.
+     * @dev     Returns the base URI for the Optimist token from the attestation.
+     * @return  string  The token URI.
+     */
+    function baseURI() public view returns (string memory) {
+        return
+            string(
+                abi.encodePacked(
+                    attestationStation.attestations(
+                        owner(),
+                        address(this),
+                        bytes32("optimist.base-uri")
+                    )
+                )
+            );
+    }
+
+    /**
+     * @notice  (Internal) Soulbound
+     * @dev     Override internal function to prevent transfers of the Optimist token.
```

Comment:
when I add the keyword override, solidity doesn't compile.   There is a test testing that this works.

Reply:
nvm figured this out.  Open zeppelin 4.8 vs 4.7 behaves differently here